packages/i18next/src/i18nextPlugin.ts

```diff
@@ -0,0 +1,134 @@
+import { match } from "@formatjs/intl-localematcher";
+import { Plugin, isNil, type Runtime } from "@squide/core";
+import type { i18n } from "i18next";
+import LanguageDetector, { type DetectorOptions } from "i18next-browser-languagedetector";
+import { i18nextInstanceRegistry } from "./i18nextInstanceRegistry.ts";
+
+export interface i18nextPluginOptions {
+    detection?: Omit<DetectorOptions, "lookupQuerystring">;
+}
+
+export function findSupportedPreferredLanguage<T extends string>(userPreferredLanguages: string[], supportedLanguages: T[]) {
+    // We don't want a fallback language here but it's a required parameter, therefore
+    // we provide a dummy value to return "undefined".
+    let result: string | undefined = match(userPreferredLanguages, supportedLanguages, "__dummy_fallback__", {
+        algorithm: "lookup"
+    });
+
+    if (result === "__dummy_fallback__") {
+        result = undefined;
+    }
+
+    if (isNil(result)) {
+        // Intl.LocaleMatcher "lookup" algorithm returns null when a prefered language is "fr" and a supported language is "fr-CA".
+        // We would prefer if it returns "fr", that's what this code is for.
+        result = supportedLanguages.find(x => {
+            return userPreferredLanguages.some(y => x.startsWith(`${y}-`));
+        });
+    }
+
+    return result;
+}
+
+export class i18nextPlugin<T extends string = string> extends Plugin {
```

Comment:
is the generic for the language used at all? i feel like since we keep casting i18nextPlugin, it can't really be enforced. Is there an advantage in the code to not have string for currentLanuage and supportedLanguages?

Reply:
Oh yeah it is:

![image](https://github.com/gsoft-inc/wl-squide/assets/794579/2d608175-1143-4583-a32e-5dcec1459d87)